File: app/ios/OffTheCloud/OffTheCloud/SecretsStore.swift
// SPDX-License-Identifier: AGPL-3.0-or-later

//
//  SecretsStore.swift
//  OffTheCloud
//
//  Created by Alonso Vidales on 8/9/25.
//


import Foundation
import Combine

final class SecretsStore: ObservableObject {
    @Published var endpoint: String
    @Published var password: String
    @Published var deviceId: String

    // Settings
    // These three auto-persist on every change (unlike endpoint/password/
    // deviceId above, which stay explicit-Save-button-only to avoid a
    // Keychain write per keystroke) — a toggle here used to only survive
    // an app relaunch if the user happened to also tap "Save Connection"/
    // "Sync Now"/"Sync All" afterward, since nothing else ever called
    // persist() for these.
    @Published var wifiOnly: Bool {
        didSet { persist() }
    }
    @Published var includeVideos: Bool {
        didSet { persist() }
    }
    @Published var downloadFromiCloud: Bool {
        didSet { persist() }
    }
    
    var isConfigured: Bool { !endpoint.isEmpty && !password.isEmpty }

    /// The endpoint as a URL the device will actually accept.
    ///
    /// People type the host and stop - "cala.off-the.cloud", or
    /// "wss://cala.off-the.cloud" - and the device's WebSocket lives at
    /// /ws, nowhere else. Without the path the bridge answers the
    /// handshake with its landing page, which URLSession reports as a
    /// bare "bad response from the server" (-1011) with nothing to say
    /// what was wrong. So the scheme and path are filled in here rather
    /// than demanded: no scheme becomes wss (a bare host is only ever a
    /// public name; the LAN case is typed with ws:// on purpose), and a
    /// missing or bare "/" path becomes /ws. Anything explicit is kept.
    static func normalizedEndpoint(_ raw: String) -> String {
        var s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !s.isEmpty else { return s }
        if !s.contains("://") {
            s = "wss://" + s
        }
        guard var parts = URLComponents(string: s) else { return s }
        if parts.path.isEmpty || parts.path == "/" {
            parts.path = "/ws"
        }
        return parts.string ?? s
    }

    /// endpoint, normalized - what every connection should dial.
    var endpointURLString: String { Self.normalizedEndpoint(endpoint) }

    // Issue #121: most devices are reached through the public bridge, where
    // the address is entirely determined by the device's name - so the
    // app asks for the name and builds the rest, and only someone with a
    // device elsewhere (their own bridge, Tailscale, the LAN) types an
    // address. These two turn a name into that address and back, so the
    // form can show whichever the stored endpoint really is.
    static let bridgeDomain = "off-the.cloud"

    static func bridgeEndpoint(forName name: String) -> String {
        "wss://" + name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() + "." + bridgeDomain + "/ws"
    }

    /// The device name if `endpoint` is a plain bridge address, else nil -
    /// which is how the form knows to open the custom field instead.
    static func bridgeName(fromEndpoint endpoint: String) -> String? {
        guard let parts = URLComponents(string: normalizedEndpoint(endpoint)),
              parts.scheme == "wss", parts.port == nil,
              parts.path == "/ws", parts.query == nil,
              let host = parts.host, host.hasSuffix("." + bridgeDomain) else { return nil }
        let name = String(host.dropLast(bridgeDomain.count + 1))
        return name.isEmpty || name.contains(".") ? nil : name
    }

    private init(endpoint: String, password: String, deviceId: String, wifiOnly: Bool, includeVideos: Bool, downloadFromiCloud: Bool) {
        self.endpoint = endpoint
        self.password = password
        self.deviceId = deviceId
        self.wifiOnly = wifiOnly
        self.includeVideos = includeVideos
        self.downloadFromiCloud = downloadFromiCloud
    }

    static func loadOrCreate() -> SecretsStore {
        let endpoint = Keychain.loadString(key: "endpoint") ?? ""
        let password = Keychain.loadString(key: "password") ?? ""
        let deviceId = Keychain.loadString(key: "device_id") ?? {
            let id = UUID().uuidString
            Keychain.saveString(key: "device_id", value: id)
            return id
        }()

        let wifiOnly = UserDefaults.standard.bool(forKey: "wifiOnly")
        let includeVideos = UserDefaults.standard.object(forKey: "includeVideos") as? Bool ?? true
        let downloadFromiCloud = UserDefaults.standard.object(forKey: "downloadFromiCloud") as? Bool ?? true

        return SecretsStore(endpoint: endpoint, password: password, deviceId: deviceId, wifiOnly: wifiOnly, includeVideos: includeVideos, downloadFromiCloud: downloadFromiCloud)
    }

    func persist() {
        Keychain.saveString(key: "endpoint", value: endpoint)
        Keychain.saveString(key: "password", value: password)
        Keychain.saveString(key: "device_id", value: deviceId)
        UserDefaults.standard.set(wifiOnly, forKey: "wifiOnly")
        UserDefaults.standard.set(includeVideos, forKey: "includeVideos")
        UserDefaults.standard.set(downloadFromiCloud, forKey: "downloadFromiCloud")
    }
}

// Tiny Keychain helper
enum Keychain {
    static func saveString(key: String, value: String) {
        let data = Data(value.utf8)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecAttrService as String: "OffTheCloud",
            kSecValueData as String: data
        ]
        let delStatus = SecItemDelete(query as CFDictionary)
        let addStatus = SecItemAdd(query as CFDictionary, nil)
        // Never log `value` here — this is also used for the account password.
        print("Keychain save key=\(key) deleteStatus=\(delStatus) addStatus=\(addStatus)")
    }

    static func loadString(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: key,
            kSecAttrService as String: "OffTheCloud",
            kSecReturnData as String: kCFBooleanTrue!,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        print("Keychain load key=\(key) status=\(status)")
        if status == errSecSuccess, let data = item as? Data {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }
}

File: app/android/app/src/main/java/cloud/offthe/otc/data/SecretsStore.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
package cloud.offthe.otc.data

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import cloud.offthe.otc.OTCApp
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.net.URI
import java.util.UUID

// Port of SecretsStore.swift. The endpoint, password and device id live in
// EncryptedSharedPreferences (the Keychain's counterpart: encrypted with a
// key in the Android Keystore); the three sync toggles in plain
// preferences, auto-persisted on change like on iOS. endpoint/password/
// deviceId only persist on an explicit persist(), so the connection form
// doesn't write a secret per keystroke.
class SecretsStore private constructor(
    endpoint: String, password: String, deviceId: String,
    wifiOnly: Boolean, includeVideos: Boolean, downloadFromCloud: Boolean,
) {
    private val _endpoint = MutableStateFlow(endpoint)
    private val _password = MutableStateFlow(password)
    private val _deviceId = MutableStateFlow(deviceId)
    private val _wifiOnly = MutableStateFlow(wifiOnly)
    private val _includeVideos = MutableStateFlow(includeVideos)
    private val _downloadFromCloud = MutableStateFlow(downloadFromCloud)

    val endpoint: StateFlow<String> = _endpoint
    val password: StateFlow<String> = _password
    val deviceId: StateFlow<String> = _deviceId
    val wifiOnly: StateFlow<Boolean> = _wifiOnly
    val includeVideos: StateFlow<Boolean> = _includeVideos
    val downloadFromCloud: StateFlow<Boolean> = _downloadFromCloud

    fun setEndpoint(v: String) { _endpoint.value = v }
    fun setPassword(v: String) { _password.value = v }
    fun setWifiOnly(v: Boolean) { _wifiOnly.value = v; persist() }
    fun setIncludeVideos(v: Boolean) { _includeVideos.value = v; persist() }
    fun setDownloadFromCloud(v: Boolean) { _downloadFromCloud.value = v; persist() }

    val isConfigured: Boolean get() = _endpoint.value.isNotEmpty() && _password.value.isNotEmpty()

    /** endpoint, normalized - what every connection should dial. */
    val endpointURLString: String get() = normalizedEndpoint(_endpoint.value)

    fun persist() {
        secure().edit()
            .putString("endpoint", _endpoint.value)
            .putString("password", _password.value)
            .putString("device_id", _deviceId.value)
            .apply()
        plain().edit()
            .putBoolean("wifiOnly", _wifiOnly.value)
            .putBoolean("includeVideos", _includeVideos.value)
            .putBoolean("downloadFromCloud", _downloadFromCloud.value)
            .apply()
    }

    companion object {
        const val bridgeDomain = "off-the.cloud"

        @Volatile private var shared: SecretsStore? = null

        /** One instance per process, loaded on first use (a Keystore round
         *  trip, so call it off the main thread the first time). */
        fun loadOrCreate(): SecretsStore = shared ?: synchronized(this) {
            shared ?: load().also { shared = it }
        }

        private fun load(): SecretsStore {
            val s = secure()
            val p = plain()
            val deviceId = s.getString("device_id", null) ?: UUID.randomUUID().toString().also {
                s.edit().putString("device_id", it).apply()
            }
            return SecretsStore(
                endpoint = s.getString("endpoint", "") ?: "",
                password = s.getString("password", "") ?: "",
                deviceId = deviceId,
                wifiOnly = p.getBoolean("wifiOnly", false),
                includeVideos = p.getBoolean("includeVideos", true),
                downloadFromCloud = p.getBoolean("downloadFromCloud", true),
            )
        }

        private fun secure(): SharedPreferences {
            val ctx = OTCApp.instance
            val key = MasterKey.Builder(ctx).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
            return EncryptedSharedPreferences.create(
                ctx, "otc_secrets", key,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
            )
        }

        private fun plain(): SharedPreferences =
            OTCApp.instance.getSharedPreferences("otc_settings", Context.MODE_PRIVATE)

        /**
         * The endpoint as a URL the device will actually accept: no scheme
         * becomes wss, a missing or bare "/" path becomes /ws. Anything
         * explicit is kept.
         */
        fun normalizedEndpoint(raw: String): String {
            var s = raw.trim()
            if (s.isEmpty()) return s
            if (!s.contains("://")) s = "wss://$s"
            return try {
                val u = URI(s)
                if (u.path.isNullOrEmpty() || u.path == "/") {
                    URI(u.scheme, u.userInfo, u.host, u.port, "/ws", u.query, u.fragment).toString()
                } else s
            } catch (e: Exception) {
                s
            }
        }

        /** Issue #121: a device name -> its bridge address. */
        fun bridgeEndpoint(forName: String): String =
            "wss://" + forName.trim().lowercase() + "." + bridgeDomain + "/ws"

        /** The device name if endpoint is a plain bridge address, else null. */
        fun bridgeName(fromEndpoint: String): String? {
            val u = try { URI(normalizedEndpoint(fromEndpoint)) } catch (e: Exception) { return null }
            val host = u.host ?: return null
            if (u.scheme != "wss" || u.port != -1 || u.path != "/ws" || u.query != null) return null
            if (!host.endsWith(".$bridgeDomain")) return null
            val name = host.dropLast(bridgeDomain.length + 1)
            return if (name.isEmpty() || name.contains(".")) null else name
        }
    }
}
